File: 3.x.x/crates-utopia/validators/src/range.rs
use crate::{Validator, ValueType};
use serde_json::Value;

#[derive(Debug, Clone)]
pub struct Range {
    min: f64,
    max: f64,
    format: &'static str,
}

impl Range {
    pub fn new(min: f64, max: f64) -> Self {
        Self {
            min,
            max,
            format: "float",
        }
    }

    pub fn integer(min: i64, max: i64) -> Self {
        Self {
            min: min as f64,
            max: max as f64,
            format: "integer",
        }
    }
}

impl Validator for Range {
    fn description(&self) -> String {
        format!(
            "Value must be a valid {} and no smaller than {} and no larger than {}",
            self.format, self.min, self.max
        )
    }

    fn value_type(&self) -> ValueType {
        if self.format == "integer" {
            ValueType::Integer
        } else {
            ValueType::Float
        }
    }

    fn is_valid(&self, value: &Value) -> bool {
        let n = match value {
            Value::Number(num) => num.as_f64(),
            Value::String(s) => s.parse::<f64>().ok(),
            _ => None,
        };
        match n {
            Some(v) => v >= self.min && v <= self.max,
            None => false,
        }
    }
}

File: packages/validators/src/Validator/Range.php
<?php

declare(strict_types=1);

namespace Utopia\Validator;

use Utopia\Validator;

/**
 * Range
 *
 * Validates that an number is in range.
 */
class Range extends Numeric
{
    public function __construct(protected int|float $min, protected int|float $max, protected string $format = self::TYPE_INTEGER) {}

    /**
     * Get Range Minimum Value
     */
    public function getMin(): int|float
    {
        return $this->min;
    }

    /**
     * Get Range Maximum Value
     */
    public function getMax(): int|float
    {
        return $this->max;
    }

    /**
     * Get Range Format
     */
    public function getFormat(): string
    {
        return $this->format;
    }

    /**
     * Get Description
     *
     * Returns validator description
     */
    #[\Override]
    public function getDescription(): string
    {
        return 'Value must be a valid range between ' . number_format($this->min) . ' and ' . number_format($this->max);
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    #[\Override]
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    #[\Override]
    public function getType(): string
    {
        return $this->format;
    }

    /**
     * Is valid
     *
     * Validation will pass when $value number is bigger or equal than $min number and lower or equal than $max.
     * Not strict, considers any valid integer to be a valid float
     * Considers infinity to be a valid integer
     */
    #[\Override]
    public function isValid(mixed $value): bool
    {
        if (!parent::isValid($value)) {
            return false;
        }

        switch ($this->format) {
            case self::TYPE_INTEGER:
                // Accept infinity as an integer
                // Since gettype(INF) === TYPE_FLOAT
                if ($value === INF || $value === -INF) {
                    break; // move to check if value is within range
                }
                $value += 0;
                if (!\is_int($value)) {
                    return false;
                }
                break;
            case self::TYPE_FLOAT:
                if (!is_numeric($value)) {
                    return false;
                }
                $value += 0.0;
                break;
            default:
                return false;
        }
        return $this->min <= $value && $this->max >= $value;
    }
}
